test/rules/invariant_boolean.dart

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `dart test/util/solo_test.dart invariant_boolean`
+
+int bar = 1;
+
+int baz = 2;
+
+int foo = 0;
+bool setting = null;
+
+void bad() {
+  if ((foo == bar && someComputation()) || (foo != bar && otherComputation())) {} // OK
+  if (foo > bar || foo == bar) {} // OK
+  if (foo > bar && someComputation() || foo == bar) {} // OK
+  if (foo > bar || foo == bar && someComputation()) {} // OK
+  if (foo < bar - baz || foo > bar) {} // OK
+  if (setting == true) {} // OK
+  if (setting != true) {} // OK
+  if (foo == bar || foo != bar) {} // LINT
```

Comment:
Added getters/fields to the tests. Right now the rule is quite complex, having it deal with side effects is simply too hard.
What about adding a rule that warns about side effects in getters? Is on my todo list now. https://github.com/dart-lang/linter/issues/293